hal/src/common/dmac/mod.rs

```diff
@@ -0,0 +1,279 @@
+//! // NOTE: buf_src and buf_dest should be either:
+//! // &'static mut T, &'static mut [T], or &'static mut [T; N] where T: BeatSize
+//! let xfer = Transfer::new(buffers, chan0, false, ());
+//! // Begin transfer with a software trigger
+//! let xfer = xfer.begin(&mut dmac, TriggerSource::DISABLE, TriggerAction::BLOCK);
+//!
+//! // Wait for transfer to complete and grab resulting buffers
+//! let (buf_src, buf_dest, chan0, _) = xfer.wait(&mut dmac);
+//! ```
+
+use modular_bitfield::prelude::*;
+
+#[cfg(feature = "min-samd51g")]
+pub use dma_controller::{BurstLength, FifoThreshold};
+pub use dma_controller::{
+    DmaController, PriorityLevel, PriorityLevelMask, RoundRobinMask, TriggerAction, TriggerSource,
+};
+use transfer::BeatSize;
+pub use transfer::{Beat, Buffer, Transfer};
+
+#[cfg(all(feature = "samd11", feature = "max-channels"))]
+#[macro_export]
+macro_rules! with_num_channels {
+    ($some_macro:ident) => {
+        $some_macro!(6);
+    };
+}
+
+#[cfg(all(feature = "samd21", feature = "max-channels"))]
+#[macro_export]
+macro_rules! with_num_channels {
+    ($some_macro:ident) => {
+        $some_macro!(12);
+    };
+}
+
+#[cfg(all(feature = "min-samd51g", feature = "max-channels"))]
+#[macro_export]
+macro_rules! with_num_channels {
+    ($some_macro:ident) => {
+        $some_macro!(32);
+    };
+}
+
+#[cfg(all(feature = "samd11", not(feature = "max-channels")))]
+#[macro_export]
+macro_rules! with_num_channels {
+    ($some_macro:ident) => {
+        $some_macro!(3);
+    };
+}
+
+#[cfg(all(feature = "samd21", not(feature = "max-channels")))]
+#[macro_export]
+macro_rules! with_num_channels {
+    ($some_macro:ident) => {
+        $some_macro!(6);
+    };
+}
+
+#[cfg(all(feature = "min-samd51g", not(feature = "max-channels")))]
+#[macro_export]
+macro_rules! with_num_channels {
+    ($some_macro:ident) => {
+        $some_macro!(16);
+    };
+}
+
+macro_rules! get {
+    ($literal:literal) => {
+        $literal
+    };
+}
+
+/// Number of DMA channels used by the driver
+pub const NUM_CHANNELS: usize = with_num_channels!(get);
+
+#[bitfield]
+#[derive(Clone, Copy)]
+#[repr(u16)]
+pub struct BlockTransferControl {
+    #[allow(dead_code)]
+    valid: bool,
+    #[allow(dead_code)]
+    evosel: B2,
+    #[allow(dead_code)]
+    blockact: B2,
+    #[skip]
+    _reserved: B3,
+    #[bits = 2]
+    #[allow(dead_code)]
+    beatsize: BeatSize,
+    #[allow(dead_code)]
+    srcinc: bool,
+    #[allow(dead_code)]
+    dstinc: bool,
+    #[allow(dead_code)]
+    stepsel: bool,
+    #[allow(dead_code)]
+    stepsize: B3,
+}
+
+// ----- DMAC SRAM registers ----- //
+/// Descriptor representing a SRAM register. Datasheet section 19.8.2
+#[derive(Clone, Copy)]
+#[repr(C, align(16))]
+#[doc(hidden)]
+pub struct DmacDescriptor {
+    btctrl: BlockTransferControl,
+    btcnt: u16,
+    srcaddr: *mut (),
+    dstaddr: *mut (),
+    descaddr: *mut DmacDescriptor,
```

Comment:
Will you ever mutate another descriptor through this pointer? Does it need to be `*mut`?

Reply:
I think these can indeed safely be `*const`